robosuite/utils/mjmod.py

```diff
@@ -1397,85 +1398,521 @@ def bitmap(self):
+                "frictionloss": np.array(self.sim.model.dof_frictionloss[dof_idx]),
+                "damping": np.array(self.sim.model.dof_damping[dof_idx]),
+            }
+
+    def restore_defaults(self):
+        """
+        Restores the default values curently saved in this modder
+        """
+        # Loop through all defaults and set the default value in sim
+        for group_defaults in (self.opt_defaults, self.body_defaults, self.geom_defaults, self.joint_defaults):
+            for name, defaults in group_defaults.items():
+                for attr, default_val in defaults.items():
+                    self.mod(name=name, attr=attr, val=default_val)
 
-    def __getattr__(self, name):
-        try:
-            opt_attr = getattr(self.opt, name)
-        except AttributeError:
-            opt_attr = None
+        # Make sure changes propagate in sim
+        self.update()
 
-        try:
-            model_attr = getattr(self.model, name)
-        except AttributeError:
-            model_attr = None
+    def randomize(self):
+        """
+        Randomizes all enabled dynamics parameters in the simulation
+        """
+        for group_defaults, group_randomizations, group_randomize_names in zip(
+                (self.opt_defaults, self.body_defaults, self.geom_defaults, self.joint_defaults),
+                (self.opt_randomizations, self.body_randomizations, self.geom_randomizations, self.joint_randomizations),
+                ([None], self.body_names, self.geom_names, self.joint_names)
+        ):
+            for name in group_randomize_names:
+                # Randomize all parameters associated with this element
+                for attr, default_val in group_defaults[name].items():
+                    val = copy.copy(default_val)
+                    settings = group_randomizations[attr]
+                    if settings["randomize"]:
+                        # Randomize accordingly, and clip the final perturbed value
+                        perturbation = np.random.rand() if type(val) in {int, float} else np.random.rand(*val.shape)
+                        perturbation = settings["perturbation"] * (-1 + 2 * perturbation)
+                        val = val + perturbation if settings["type"] == "size" else val * (1.0 + perturbation)
+                        val = np.clip(val, *settings["clip"])
+                    # Modify this value
+                    self.mod(name=name, attr=attr, val=val)
+
+        # Make sure changes propagate in sim
+        self.update()
 
-        ret = opt_attr if opt_attr is not None else model_attr
-        if callable(ret):
-            def r(*args):
-                return ret(*args)
-            return r
+    def update_sim(self, sim):
+        """
+        In addition to super method, reset all internal values
+
+        Args:
+            sim (MjSim): MjSim object
+        """
+        super().update_sim(sim=sim)
+        self.save_defaults()
```

Comment:
This doesn't seem to restore the defaults, but just saves them - which is in conflict with the docstring - what's the expected behavior?

Reply:
Hm, that is a good point, perhaps the wording was unclear. By "reset", I mean that the internal default values are updated (via the `save_defaults()` method), since I assumed that if you're calling this method, the old `sim` object was stale and there's now a new simulation instance whose dynamics values need to be kept track of. Is there a better way to convey this?

I don't think it would make sense to actually restore the defaults in sim using this method -- for example, if you're reloading from an XML, and want to update the sim for this modder, it wouldn't necessarily make sense to override the values from the loaded XML, right?

Reply:
sure - when I think of "reset", I think about resetting to old values, not clearing them and saving the new defaults - so it would be nice to clarify in docstring

Reply:
Done!